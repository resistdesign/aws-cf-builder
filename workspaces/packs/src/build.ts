import { createResourcePack } from '@aws-cf-builder/utils';

export const addBuildPipeline = createResourcePack(({ id }: { id: string }) => {
  return {
    Resources: {
      CodeBuildRole: {
        Type: 'AWS::IAM::Role',
        Properties: {
          AssumeRolePolicyDocument: {
            Statement: [
              {
                Effect: 'Allow',
                Principal: {
                  Service: ['codebuild.amazonaws.com'],
                },
                Action: ['sts:AssumeRole'],
              },
            ],
          },
          Path: '/',
          Policies: [
            {
              PolicyName: 'codebuild-service',
              PolicyDocument: {
                Statement: [
                  {
                    Effect: 'Allow',
                    Action: '*',
                    Resource: '*',
                  },
                ],
                Version: '2012-10-17T00:00:00.000Z',
              },
            },
          ],
        },
      },
      CodePipelineRole: {
        Type: 'AWS::IAM::Role',
        Properties: {
          AssumeRolePolicyDocument: {
            Statement: [
              {
                Effect: 'Allow',
                Principal: {
                  Service: ['codepipeline.amazonaws.com'],
                },
                Action: ['sts:AssumeRole'],
              },
            ],
          },
          Path: '/',
          Policies: [
            {
              PolicyName: 'codepipeline-service',
              PolicyDocument: {
                Statement: [
                  {
                    Action: ['codebuild:*'],
                    Resource: '*',
                    Effect: 'Allow',
                  },
                  {
                    Action: ['s3:GetObject', 's3:GetObjectVersion', 's3:GetBucketVersioning'],
                    Resource: '*',
                    Effect: 'Allow',
                  },
                  {
                    Action: ['s3:PutObject'],
                    Resource: ['arn:aws:s3:::codepipeline*'],
                    Effect: 'Allow',
                  },
                  {
                    Action: ['s3:*', 'cloudformation:*', 'iam:PassRole'],
                    Resource: '*',
                    Effect: 'Allow',
                  },
                ],
                Version: '2012-10-17T00:00:00.000Z',
              },
            },
          ],
        },
      },
      PipelineBucket: {
        Type: 'AWS::S3::Bucket',
        DeletionPolicy: 'Delete',
        Properties: {
          BucketEncryption: {
            ServerSideEncryptionConfiguration: [
              {
                ServerSideEncryptionByDefault: {
                  SSEAlgorithm: 'AES256',
                },
              },
            ],
          },
          PublicAccessBlockConfiguration: {
            BlockPublicAcls: true,
            BlockPublicPolicy: true,
            IgnorePublicAcls: true,
            RestrictPublicBuckets: true,
          },
        },
      },
      CodeBuildDeploySite: {
        Type: 'AWS::CodeBuild::Project',
        DependsOn: ['CodeBuildRole', 'S3', 'CloudFront'],
        Properties: {
          Name: null,
          Description: 'Deploy site to S3',
          ServiceRole: null,
          Artifacts: {
            Type: 'CODEPIPELINE',
          },
          Environment: {
            Type: null,
            ComputeType: null,
            Image: null,
            EnvironmentVariables: [
              {
                Name: 'APP_DOMAIN',
                Type: 'PLAINTEXT',
                Value: null,
              },
              {
                Name: 'API_DOMAIN',
                Type: 'PLAINTEXT',
                Value: null,
              },
              {
                Name: 'USER_POOL_CLIENT_ID',
                Type: 'PLAINTEXT',
                Value: null,
              },
              {
                Name: 'USER_POOL_CLIENT_COOKIE_DOMAIN',
                Type: 'PLAINTEXT',
                Value: null,
              },
              {
                Name: 'USER_POOL_CLIENT_DOMAIN',
                Type: 'PLAINTEXT',
                Value: null,
              },
              {
                Name: 'REGION',
                Type: 'PLAINTEXT',
                Value: null,
              },
              {
                Name: 'OAUTH_CONFIG_DOMAIN',
                Type: 'PLAINTEXT',
                Value: {
                  'Fn::Sub': [
                    'https://cognito-idp.${UserPoolRegion}.amazonaws.com/${UserPoolID}',
                    {
                      UserPoolRegion: null,
                      UserPoolID: null,
                    },
                  ],
                },
              },
            ],
          },
          Source: {
            Type: 'CODEPIPELINE',
            BuildSpec: {
              'Fn::Sub': [
                'version: 0.2\nphases:\n  build:\n    commands:\n      - ${ConfigCommand}\n      - "npm i yarn && npx yarn && API_BASE_URL=\'https://${FullAPIDomainName}\' NODE_ENV=${EnvironmentType} npx yarn workspace @er-candidates/app build"\n  post_build:\n    commands:\n      - aws s3 cp --recursive --acl public-read ${OutputDirectory} s3://${FullDomainName}/\n      - aws cloudfront create-invalidation --distribution-id "${DistributionId}" --paths "/*"\n',
                {
                  DistributionId: null,
                  FullDomainName: null,
                  FullAPIDomainName: null,
                  OutputDirectory: null,
                },
              ],
            },
          },
          TimeoutInMinutes: 10,
        },
      },
      Pipeline: {
        Type: 'AWS::CodePipeline::Pipeline',
        DependsOn: 'CodeBuildDeploySite',
        Properties: {
          RoleArn: null,
          Stages: [
            {
              Name: 'Acquire-Source',
              Actions: [
                {
                  InputArtifacts: [],
                  Name: 'Source',
                  ActionTypeId: {
                    Category: 'Source',
                    Owner: 'ThirdParty',
                    Version: '1',
                    Provider: 'GitHub',
                  },
                  OutputArtifacts: [
                    {
                      Name: 'SourceOutput',
                    },
                  ],
                  Configuration: {
                    Owner: null,
                    Repo: null,
                    Branch: null,
                    OAuthToken: null,
                  },
                  RunOrder: 1,
                },
              ],
            },
            {
              Name: 'Build-And-Deploy',
              Actions: [
                {
                  Name: 'Artifact',
                  ActionTypeId: {
                    Category: 'Build',
                    Owner: 'AWS',
                    Version: '1',
                    Provider: 'CodeBuild',
                  },
                  InputArtifacts: [
                    {
                      Name: 'SourceOutput',
                    },
                  ],
                  OutputArtifacts: [
                    {
                      Name: 'DeployOutput',
                    },
                  ],
                  Configuration: {
                    ProjectName: null,
                  },
                  RunOrder: 1,
                },
              ],
            },
          ],
          ArtifactStore: {
            Type: 'S3',
            Location: null,
          },
        },
      },
    },
  };
});
