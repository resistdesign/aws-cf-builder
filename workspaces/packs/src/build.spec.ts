import { addBuildPipeline } from './build';

describe('Build', () => {
  describe('addBuildPipeline', () => {
    test('should add a build pipeline', () => {
      const cft = addBuildPipeline(
        {
          id: 'APIBuild',
          buildSpec: 'echo "build spec"',
          repoConfig: {
            owner: 'owner',
            repo: 'repo',
            branch: 'main',
            oauthToken: 'oauth-token',
          },
        },
        { AWSTemplateFormatVersion: '2010-09-09' },
      );

      expect(cft).toMatchSnapshot();
    });
  });
});
