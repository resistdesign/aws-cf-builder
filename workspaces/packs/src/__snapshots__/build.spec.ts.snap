// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build addBuildPipeline should add a build pipeline 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Metadata": Object {
    "AWS::CloudFormation::Interface": Object {
      "ParameterGroups": Array [
        Object {
          "Label": Object {
            "default": "APIBuild Parameters",
          },
          "Parameters": Array [
            "APIBuildGitHubUser",
            "APIBuildGitHubRepo",
            "APIBuildGitHubBranch",
            "APIBuildGitHubToken",
          ],
        },
      ],
      "ParameterLabels": Object {
        "APIBuildGitHubBranch": Object {
          "default": "APIBuild GitHub Branch",
        },
        "APIBuildGitHubRepo": Object {
          "default": "APIBuild GitHub Repo",
        },
        "APIBuildGitHubToken": Object {
          "default": "APIBuild GitHub Token",
        },
        "APIBuildGitHubUser": Object {
          "default": "APIBuild GitHub User",
        },
      },
    },
  },
  "Parameters": Object {
    "APIBuildGitHubBranch": Object {
      "Default": "master",
      "Description": "The name of the branch to be built",
      "Type": "String",
    },
    "APIBuildGitHubRepo": Object {
      "Description": "The name of the GitHub repository",
      "Type": "String",
    },
    "APIBuildGitHubToken": Object {
      "Description": "A GitHub Access Token with \`repo\` permissions (https://github.com/settings/tokens)",
      "NoEcho": true,
      "Type": "String",
    },
    "APIBuildGitHubUser": Object {
      "Description": "The GitHub User that owns the repository",
      "Type": "String",
    },
  },
  "Resources": Object {
    "APIBuildCodeBuildAndDeploy": Object {
      "DependsOn": undefined,
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "Description": "Deploy site to S3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": undefined,
          "Image": "aws/codebuild/nodejs:10.14.1",
          "Type": "LINUX_CONTAINER",
        },
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-APIBuildCodeBuildAndDeploy",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "APIBuildCodeBuildRole",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "echo \\"build spec\\"",
          "Type": "CODEPIPELINE",
        },
        "TimeoutInMinutes": 10,
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "APIBuildCodeBuildRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "codebuild.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "*",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "codebuild-service",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "APIBuildCodePipelineRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "codepipeline.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "codebuild:*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "s3:GetObject",
                    "s3:GetObjectVersion",
                    "s3:GetBucketVersioning",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "s3:PutObject",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    "arn:aws:s3:::codepipeline*",
                  ],
                },
                Object {
                  "Action": Array [
                    "s3:*",
                    "cloudformation:*",
                    "iam:PassRole",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "codepipeline-service",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "APIBuildPipeline": Object {
      "DependsOn": "APIBuildCodeBuildAndDeploy",
      "Properties": Object {
        "ArtifactStore": Object {
          "Location": Object {
            "Ref": "APIBuildPipelineBucket",
          },
          "Type": "S3",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "APIBuildCodePipelineRole",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": Object {
                    "Ref": "APIBuildGitHubBranch",
                  },
                  "OAuthToken": Object {
                    "Ref": "APIBuildGitHubToken",
                  },
                  "Owner": Object {
                    "Ref": "APIBuildGitHubUser",
                  },
                  "Repo": Object {
                    "Ref": "APIBuildGitHubRepo",
                  },
                },
                "InputArtifacts": Array [],
                "Name": "Source",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "SourceOutput",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Acquire-Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "APIBuildCodeBuildAndDeploy",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "SourceOutput",
                  },
                ],
                "Name": "Artifact",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "DeployOutput",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Build-And-Deploy",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "APIBuildPipelineBucket": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
    },
  },
}
`;
